test/tests/functional/pbs_hook_management.py

```diff
@@ -0,0 +1,151 @@
+# coding: utf-8
+
+# Copyright (C) 1994-2019 Altair Engineering, Inc.
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+import os
+import time
+import socket
+import textwrap
+import datetime
+from tests.functional import *
+from tests.functional import MGR_CMD_SET
+from tests.functional import SERVER
+from tests.functional import ATTR_h
+from tests.functional import TEST_USER
+from tests.functional import Job
+from tests.functional import JOB
+
+
+def get_hook_body(hook_msg):
+    hook_body = """
+    import pbs
+    e = pbs.event()
+    m = e.management
+    pbs.logmsg(pbs.LOG_DEBUG, '%s')
+    """ % hook_msg
+    hook_body = textwrap.dedent(hook_body)
+    return hook_body
```

Comment:
I would like your ptl test to be expanded some more. There are things in your design doc need to be exercised. 

- The pbs.event().management.* attribute values need to be shown and verified as stated in the design:
    

> pbs.event().management: _management
>         cmd: integer value of the server management operation, constants defined in pbs.MGR_CMD_*, translation to text is found in pbs.REVERSE_MGR_CMDS
>         objname: name of the object, such as the hook name.
>         objtype: integer value of the server management operation type, constants defined in pbs.MGR_OBJ_*, translation to text is found in pbs.REVERSE_MGR_OBJS
>         request_time: time of the request in seconds since epoch
>         reply_text: text returned to the user via the server management operation
>         reply_code: code returned to the user via the server management operation
>         reply_choice: request choice, constants defined in pbs.BRP_CHOICE_*, translation to text is found in pbs.REVERSE_BRP_CHOICES
>         reply_auxcode: Auxiliary error code
>         attribs: List of server_attribute class objects
>     pbs.event().management.rq_attrs: List[_server_attribute]
>         name: name of the attribute
>         resource: resource of the attribute
>         value: value of the attribute
>         op: integer value of the attribute operation, constants defined in pbs.BATCH_OP_*, translation to text is found in pbs.REVERSE_BATCH_OPS
>         flags: integer value of the attribute flags, constants defined in pbs.ATR_VFLAG_*, translation to text is found in pbs.REVERSE_ATR_VFLAGS
>         sisters: list of co-resources for the attribute

As an example, I created this management hook:
`import pbs`
`import pbs`
`e = pbs.event()`
`m = e.management`
`pbs.logmsg(pbs.LOG_DEBUG, "executed management hook order 2")`
`pbs.logmsg(pbs.LOG_DEBUG, "cmd=%s" % (pbs.REVERSE_MGR_CMDS[m.cmd]))`
`pbs.logmsg(pbs.LOG_DEBUG, "objtype=%s" % (pbs.REVERSE_MGR_OBJS[m.objtype]))`
`pbs.logmsg(pbs.LOG_DEBUG, "reply_choice=%s" % (pbs.REVERSE_BRP_CHOICES[m.reply_choice]))`
`for a in m.attribs:`
 `\   pbs.logmsg(pbs.LOG_DEBUG, "%s" % (a))`

When I ran qmgr -c 'set hook qmgr enabled=true', the management hook executed and produced this output:
03/13/2020 02:24:31.334703;0006;Server@corretja;Hook;Server@corretja;executed management hook order 2
03/13/2020 02:24:31.334709;0006;Server@corretja;Hook;Server@corretja;cmd=MGR_CMD_SET
03/13/2020 02:24:31.334713;0006;Server@corretja;Hook;Server@corretja;objtype=MGR_OBJ_HOOK
03/13/2020 02:24:31.334717;0006;Server@corretja;Hook;Server@corretja;reply_choice=BRP_CHOICE_NULL
03/13/2020 02:24:31.334731;0006;Server@corretja;Hook;Server@corretja;a=name=enabled:resource=None:value=true:op=0:flags=0:sisters=[]`

- You should also have test cases for the following scenarios as mentioned in the design:

> Keeping with the hook design, if one management hook is rejected, the other management hooks with a higher order value will not run.
> 
> If the management hook script encounters an unexpected error causing an unhandled exception, or times out due to the hook's alarm setting, the hook will act similar to a pbs.event().reject().

- Finally, when I added in my test management hook the call "pbs.event().reject("bad hook"), I did not see the message "bad hook" in the server_logs. Whatever message is passed to the reject() call should show up in the server_logs.

Reply:
I am adding the tests now.  I have encountered two errors in server_process_hooks that need to be looked at before I can continue.